### ISSUE TITLE
Very strange issue with `writer.close` (silently closes entire node process)

### ISSUE THREAD

**Issue by piboistudios** [ipld/js-car#140]
Using `@ipld/car` version 5.1.1

This simple code snippet is self explanatory, and happens as the `console.log`s suggest... though you would expect that it doesn't:
```
import { CarWriter } from "@ipld/car";

const main = async () => {
    console.log('wtf');
    const { writer, out } = await CarWriter.create();
    console.log('lol');
    await writer.close();
    console.log('this never prints'); // this really never prints
}

main()
    .then(() => {
        console.log('done'); // this also never prints
    })
    .catch(e => {
        console.error('ERROR:', e); // no errors either
    })
 ```
REPL to prove I'm not lying: 
https://replit.com/@GabrielSpeed/ToughRadiantMarketing#index.js
(Also happens with BunJS https://replit.com/@GabrielSpeed/CreepyGigaGnudebugger#index.ts)

ETA: this happens even if the CAR is created with roots, and also even if the CAR is created with roots and the block containing the root is actually in the CAR.

Basically this happens no matter what state the `CarWriter` is in. Not sure why or how, but I will tell you the bug is going to be incredibly hard to solve, because it prints no error message or anything at any level.

Seems to have something to do with methods on the `_encoder` of the `CarWriter`. Something about the way the `_mutex` and `_encoder` play together (and that IteratorChannel_Writer class) is not working out well, and seems to be a problem similar to this: https://stackoverflow.com/questions/64897131/async-for-loop-causes-the-function-to-silently-terminate

**Comment by piboistudios** [comment]
Also, what was interesting is that if you use `writer._encoder.close()` directly, it works fine... for a while.

It will still eventually close the process if you run enough operations concurrently (on separate writers). Definitely something going on with the encoder.

**Comment by piboistudios** [comment]
So, apparently, the CAR Writer only works if the `AsyncIterable<UInt8Array>` is converted to a node Stream at some point (e.g. via `Readable.from(out).pipe(fs.createWriteStream('test' + '.car'))`).

If you try to consume the AsyncIterable<UInt8Array> directly, the entire application will crash.

This is quite problematic considering `ipfs.dag.import(source)` takes a `AsyncIterable<UInt8Array>` for the `source` parameter....

I now basically have to write the CAR to a file, then immediately read the file into memory and delete it in order for the code to work....

**Comment by piboistudios** [comment]
Yeah... there is something quite unstable about how this library is writing CARs.

I have, at this point, had to wrap the code dealing with the `CarWriter` with a call from `async.retry` to have it retry a few times as it would randomly fail with the reasoning being `Unexpected end of data`

The code finally runs after several monkey patches, and at that, it is significantly slower because now it sometimes has to rewrite CARs randomly..

**Comment by piboistudios** [comment]
The `CarBlockIterator`'s static method `fromIterable` also produces an iterable that appears to just hang.

Causing `ipfs.dag.import` to likewise, just hang: https://github.com/ipfs/js-ipfs/blob/b64d4af034f27aa7204e57e6a79f95461095502c/packages/ipfs-core/src/components/dag/import.js#L92

This is using the CAR produced by the `CarWriter` class.

The CAR file: https://filebin.net/s55v8pfj6b52q0ia

**Comment by piboistudios** [comment]
It seems like all of the issues with this library, for reading at least, stem from the fact that the reader attempts to allow concurrent `put`s; there is a _lot_ of code written to accommodate this behavior... (and when you actually use it, it just produces bugs...)

Very basic use cases of the library are broken... for both reading and writing. No offense, but I'm not sure how there is a suite of unit tests and almost nothing works in the most basic (and important) cases.. At this point I am just going to write my own utility for dealing with CARs...

**Comment by rvagg** [comment]
I hope the venting was therapeutic. But a word of advice - if you want help from authors of open source libraries, abuse and criticism that isn't framed constructively rarely leads to positive outcomes.

This looks like simply a usage issue here, or a misunderstanding about what is going on.

```
    const { writer, out } = await CarWriter.create();
    console.log('lol');
    await writer.close();
```

What are you expecting to happen here? The CAR writer is trying to write a header _somewhere_ but you're not consuming it. `out` needs to go somewhere, it the whole construct can't be closed and cleaned up until that's flushed.

If you need to steam I'd also recommend using `pipeline()` rather than `pipe()` directly so you get errors handled properly.

Set up your `out` consumption before you write into `writer`. You also need to take care of the asynchronous nature of the flow, you can't manage both of these things in the same flow (`await`ing the `out` chunks while also `await`ing the `writer` writes, you'll just get stuck).

**Comment by piboistudios** [comment]
> What are you expecting to happen here?

```js
    .catch(e => {
        console.error('ERROR:', e); // no errors either
    })
```
Really anything but the program silently closing.

> Set up your out consumption before you write into writer. You also need to take care of the asynchronous nature of the flow, you can't manage both of these things in the same flow (awaiting the out chunks while also awaiting the writer writes, you'll just get stuck).

How is that at all intuitive or obvious from any of the documentation, examples or test cases the library has? 

The documentation presents `await`ing `write`s as an option, not a requirement if you are awaiting downstream.

The way you are intending this library to be used is not forthcoming at all. Hopefully that point is not lost on you solely because of my annoyance with the way this library is designed.
